### PR TITLE
fix: treat certain US numbers as invalid

### PR DESCRIPTION
**Pull request by barros001**
### Details

We have a special case of US phone numbers that are technically valid but we want to treat them as invalid.

### Fixed Issues

$ https://github.com/Expensify/App/issues/28492
PROPOSAL: https://github.com/Expensify/App/issues/28492#issuecomment-1785499023

### Tests

1. Click the search icon
2. Type +15005550009
3. Verify a phone number suggestion is displayed
4. Add an extra 1 after +1 (+15005550009)
5. Verify invalid phone number message is displayed
 
- [x] Verify that no errors appear in the JS console

### Offline tests

1. Click the search icon
2. Type +15005550009
3. Verify a phone number suggestion is displayed
4. Add an extra 1 after +1 (+15005550009)
5. Verify invalid phone number message is displayed

### QA Steps

1. Click the search icon
2. Type +15005550009
3. Verify a phone number suggestion is displayed
4. Add an extra 1 after +1 (+15005550009)
5. Verify invalid phone number message is displayed

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android_native.webm](https://github.com/Expensify/App/assets/765936/66ba01d2-38a4-493b-b0ea-657ea7b77a5a)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[android_web.webm](https://github.com/Expensify/App/assets/765936/4470de08-845d-4dbf-b753-20a3eac1d8f8)

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/765936/53b887ce-f111-4a6b-9789-301ed500777d

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/765936/9b35e203-c1c4-4a55-ba42-a895ae8b4f39

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/765936/d571c30c-9b02-4922-bbbc-16c2945a087b

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/765936/f0069233-869b-4051-b676-98cc3a65f918

</details>

### CODE REVIEW

File: tests/unit/PhoneNumberTest.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import {parsePhoneNumber} from '@libs/PhoneNumber';
+
+describe('PhoneNumber', () => {
```
[the remark this answers is not in the document]
Comment (barros001) [reply tests/unit/PhoneNumberTest.js:3]: @aimane-chnaif added a few more test numbers

File: src/libs/PhoneNumber.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            international: `${countryCode} ${phoneNumberWithoutCountryCode}`,
+            national: phoneNumberWithoutCountryCode,
+            rfc3966: `tel:${countryCode}-${phoneNumberWithoutCountryCode}`,
+            significant: phoneNumberWithoutCountryCode,
+        },
+    } as ParsedPhoneNumberInvalid;
+}
+
+// eslint-disable-next-line import/prefer-default-export
+export {parsePhoneNumber};
```
Comment (barros001) [inline src/libs/PhoneNumber.ts:42]: I wasn't very happy with the `export {*} from 'awesome-phone-number'` part. I I think it was unnecessary and added complexity to it. I'm now only exporting `parsePhoneNumber` and I updated the lint rule to only restrict this function from `awesome-phone-number` (we are not using anything else throughout the entire codebase anyways).

**Review by aimane-chnaif — COMMENTED** [review]
2 bugs in this video:
- error text glitch
- no +11 validation when there are spaces between letters

https://github.com/Expensify/App/assets/96077027/fec6fe8c-2e1c-4f9a-8030-be9eba4ea7a3

**Comment by barros001** [comment]
> 2 bugs in this video:
> 
> * error text glitch
> * no +11 validation when there are spaces between letters
> 
>  bug.mov

I believe the glitch is unrelated to this fix as it’s also happening on staging. I’ll try to find the cause and possibly fix it. 

I’ll look into the validation issue and report back.

**Comment by barros001** [comment]
> no +11 validation when there are spaces between letters

this is also not related to this fix. You can see this happening on staging with the following examples:

+15555555555 (try adding a space after the +)
+1 (555) 555-5555

this is because of the following line:

https://github.com/Expensify/App/blob/0a9f7d42977daf3faabbb4d33731e28a3e172407/src/libs/OptionsListUtils.js#L1631

When a telephone is not valid (possible is false) it will check if the search string is only numbers and +, and if so then it will display the error message related to phone numbers. Because space (and all other non numeric chars) is not part of it, it will think the search string is not a phone number. We could add spaces, () and - to the list as well to address that (might not be as simple because it would then show the phone error if you type something like `()` - basically any string with only ()- without numbers.)

What are your thoughts? Should we try and address both issues you pointed out as part of this PR even though they are not really related? I’m fine either way, just want to make sure to follow the correct procedure here.

**Comment by barros001** [comment]
> > no +11 validation when there are spaces between letters
> 
> this is also not related to this fix. You can see this happening on staging with the following examples:
> 
> +15555555555 (try adding a space after the +) +1 (555) 555-5555
> 
> this is because of the following line:
> 
> https://github.com/Expensify/App/blob/0a9f7d42977daf3faabbb4d33731e28a3e172407/src/libs/OptionsListUtils.js#L1631
> 
> When a telephone is not valid (possible is false) it will check if the search string is only numbers and +, and if so then it will display the error message related to phone numbers. Because space (and all other non numeric chars) is not part of it, it will think the search string is not a phone number. We could add spaces, () and - to the list as well to address that (might not be as simple because it would then show the phone error if you type something like `()` - basically any string with only ()- without numbers.)
> 
> What are your thoughts? Should we try and address both issues you pointed out as part of this PR even though they are not really related? I’m fine either way, just want to make sure to follow the correct procedure here.

As far as the above, we can replace `DIGITS_AND_PLUS` with the regex below:
```diff
-        DIGITS_AND_PLUS: /^\+?[0-9]*$/,
+        PHONE_NUMBER_CHARS: /^\+?(?=.*[0-9])([0-9() -]+)$/,
```

Basically it will expand the list of characters to include `(`, `)`, `-` and space, while requiring at least one number. This is a more comprehensive way of guessing if the user is trying to type a phone number of just text.

I haven't made the change yet, just wanted to hear your thoughts first.

**Comment by barros001** [comment]
On main, +1 2 5 00 will say “no results” while +12500 will show the invalid phone number error.

File: tests/unit/PhoneNumberTest.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (barros001) [reply tests/unit/PhoneNumberTest.js:3]: Sure. I’ll be spotty for the next 24 hours due to long haul traveling, but I’ll add it as soon as I can.

[the remark this answers is not in the document]
Comment (barros001) [reply tests/unit/PhoneNumberTest.js:3]: @aimane-chnaif added a few extra numbers with spaces between every number including country selector. Let me know if this is good enough.

**Review by aimane-chnaif — APPROVED** [review]
LGTM 🎉

**Comment by barros001** [comment]
> @barros001 there's conflict

@aimane-chnaif conflict resolved. We should try and merge this PR quickly because conflicts tend to happen the longer we wait due to the nature of the changes.

**Review by madmax330 — APPROVED** [review]

**Comment by mountiny** [comment]
@barros001  @aimane-chnaif   This was reverted as the additional phone check on the Search page has been causing the app to hang/ crash for Account Manager users who have thousands of people with sms number in their contacts

With the next attempt, lets make an Adhoc build and ask internal users like @conorpendergrast or @muttmuure to test the build search page performance

**Comment by mountiny** [comment]
No problem, this was super hard to catch because only some of the largest heavy users internally had this issue

**Comment by mountiny** [comment]
@barros001 I must admit that there was not a clear reasoning due to the fact this was hard to experience locally, we ahve reverted two PRs, this one and https://github.com/Expensify/App/pull/31010 so it could be that main problem was actually coming from the other PR.

Nevertheless, you are right the search page needs optimizing and there is couple PRs already in works as well as Reassure tests so hopefully we will get into a better place in a week or so

**Pull request merged**